com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs

```diff
@@ -504,6 +509,183 @@ internal void InvokeBehaviourOnGainedOwnership()
+                    }
+                }
+            }
+
+            m_IsReparented = true;
+            ApplyNetworkParenting();
+        }
+
+        internal static void WriteNetworkParenting(NetworkWriter writer, bool isReparented, ulong? latestParent)
+        {
+            writer.WriteBool(isReparented);
+            if (isReparented)
+            {
+                var isLatestParentSet = latestParent != null && latestParent.HasValue;
+                writer.WriteBool(isLatestParentSet);
+                if (isLatestParentSet)
+                {
+                    writer.WriteUInt64Packed(latestParent.Value);
+                }
+            }
+        }
+
+        internal static (bool IsReparented, ulong? LatestParent) ReadNetworkParenting(NetworkReader reader)
+        {
+            ulong? latestParent = null;
+            bool isReparented = reader.ReadBool();
+            if (isReparented)
+            {
+                var isLatestParentSet = reader.ReadBool();
+                if (isLatestParentSet)
+                {
+                    latestParent = reader.ReadUInt64Packed();
+                }
+            }
+
+            return (isReparented, latestParent);
+        }
+
+        internal (bool IsReparented, ulong? LatestParent) GetNetworkParenting() => (m_IsReparented, m_LatestParent);
+
+        internal void SetNetworkParenting(bool isReparented, ulong? latestParent)
+        {
+            m_IsReparented = isReparented;
+            m_LatestParent = latestParent;
+        }
+
+        internal static HashSet<NetworkObject> OrphanChildren = new HashSet<NetworkObject>();
+
+        internal bool ApplyNetworkParenting()
+        {
+            if (!IsSpawned)
+            {
+                return false;
+            }
+
+            if (!m_IsReparented)
+            {
+                return true;
+            }
+
+            if (m_LatestParent == null || !m_LatestParent.HasValue)
+            {
+                m_CachedParent = null;
+                transform.parent = null;
+                return true;
+            }
+
+            if (!NetworkManager.SpawnManager.SpawnedObjects.ContainsKey(m_LatestParent.Value))
+            {
+                if (OrphanChildren.Add(this))
+                {
+                    if (NetworkLog.CurrentLogLevel <= LogLevel.Normal)
+                    {
+                        NetworkLog.LogWarning($"{nameof(NetworkObject)} ({name}) cannot find its parent, added to {nameof(OrphanChildren)} set");
+                    }
+                }
+                return false;
+            }
+
+            var parentObject = NetworkManager.SpawnManager.SpawnedObjects[m_LatestParent.Value];
+
+            m_CachedParent = parentObject.transform;
+            transform.parent = parentObject.transform;
+
+            return true;
+        }
+
+        internal static void CheckOrphanChildren()
```

Comment:
Wonder if any of the related code could leverage from OnTransformChildrenChanged...

Reply:
we're keeping a list of so-called `OrphanChildren` `NetworkObject`s because at the time we initialize/spawn `NetworkObject` locally, we might not have its parent replicated from the other side (I tried to explain in my internal Slack video).
for instance, if we're spawning `NetworkObject` 5 and its parent is 10, what should happen if we do not have 10 yet? let's say 10 is on the way to be replicated in a few frames and we could fix that parent-child relationship later.
that's what we're trying to achieve here — if you couldn't find your parent, we put you into a list and everytime we spawn another `NetworkObject` locally due to replication, we quickly iterate over `OrphanChildren` and see if we can reparent/adopt one.

Reply:
The above would make a really cool in-code comment :)

Reply:
I think simple/elegant code is a thousand times better than the best inline comment. what would make you confused in this context without the comment above for instance?

Reply:
At first read, I didn't know what were orphan children, how come that could be a possibility at all. I didn't watch your video yet and you can't assume everyone reading this code in the future will either :/

Reply:
you don't have to watch the video. `OrphanChildren` appears in 2 methods: `ApplyNetworkParenting()` and `CheckOrphanChildren()` which are on top of each other. I don't understand how inline comments would make it any better than what it is currently.

Reply:
oh it's not about the what, it's about the why. **why** is there orphan children. which you explain super clearly in your comment above.

Reply:
```cs
NetworkLog.LogWarning($"{nameof(NetworkObject)} ({name}) cannot find its parent, added to {nameof(OrphanChildren)} set");
```

because `NetworkObject` can't find its parent.

the scenario I was talking about above is specific and does not cover all cases when an object's parent might be missing. that warning log explains what is happening and tells you what it is going to do next. so, why is the logic implemented right there, not an example scenario. logic is implemented to make it robust against unparented objects, not to make it responsive to the example given out there, it just happens to be.

Reply:
When I read the OrphanChildren variable, my first interpretation "this is where lost objects go", (e.g. an error condition) which I guess is kinda correct, but having read the above explanation realized it was "this is where unresolved objects go, which is a normal condition due to ordering issues".  So I would second putting your nicely-written explanation in the code pls

Reply:
sure, will try to comment inline.

Reply:
done.